### ISSUE TITLE
db: Finer-grained Time-Bound Iteration (TBI)

### ISSUE THREAD

**Issue by sumeerbhola** [cockroachdb/pebble#1190]
In the CockroachDB context, we have used the `TablePropertyCollector` and `IterOptions.TableFilter` hook to implement a time-bound Iterator (TBI) which only iterates over sstables that contain data that intersects with a  a (Tstart, Tend] time interval of the MVCC timestamps. This is currently used for more efficient export requests (incremental backups) and planned to be used for rangefeed catchup scans, both of which need to look at only recent data wrt timestamps.
Another use case that has come up is to find expired data when enforcing row-level TTLs, where (Tstart, Tend] will be in the distant past (imagine expiring 1 year old data, and doing this expiry every day, so we would be effectively looking for (now-366d, now-365d]. Since TBI filters at the granularity of sstables, and L6 contains 90% of the data, the expectation is that the current TBI will be ineffective for this use case.

In 21.1 we introduced read-triggered compactions, which use spare compaction resources and knowledge of read amplification encountered by reads to prioritize compacting down sstables in parts of the key space that would most benefit from reduced read amplification. The assumption here is that compacted data is never harmful for reads. The above TBI optimization violates this assumption -- if all data is compacted down to L6, and the keys that have been updated in (Tstart, Tend] are sparse, but evenly distributed across the key space, every sstable will be included in the TBI. This bad interaction between TBI and read-triggered compactions has been observed in practice in CockroachDB's telemetry cluster. TBI is also fragile under manual compactions, which we sometimes do when restoring a poorly shaped LSM to health.

We cannot truly lay the blame for TBI's fragility to read-triggered compactions: using (spare) resources to reduce read amplification should always be a worthy goal of the system, and anything that goes against that should be fixed. One approach would be to explicitly keep older data in separate ssblocks. In addition to being a very significant change to how we layout data, it also means more merging when using regular iterators. The following documents a simpler, high-level approach derived from a detailed proposal by @jbowens. Note that all such approaches require Pebble to have some knowledge of timestamps (unlike the current TBI approach).

**Assumption:** Filtering at the granularity of sstables is too coarse, but filtering at the granularity of ssblocks should be sufficient if most rows in an sstable are not relevant to the TBI time interval. This assumption needs to be validated with normal workloads. In the worst case we could get unlucky and need 1 row from each ssblock. Note that we would continue to also filter by sstable.

**Approach:**
- Maintain (in a time index in the sstable) what is conceptually a set containing [min,max] timestamps for each ssblock.
- When seeking to block i, based on the key index, look for the first block >= i whose timestamp interval overlaps with the TBI interval, and seek to that block.
  - When a seek is not able to use a Next optimization, both the key index block, and data block is seeked. We are not able to reduce the number of such seeks with this finer-grained TBI optimization. But note that we will continue to use the coarse-grained optimization too, and when the coarse-grained optimization is ineffective, it is often due to more compactions, which means we have fewer files and therefore fewer seeks.
  - Most of the use cases involve few seeks and mostly Next/Prev since they are scanning over a whole CockroachDB range, so the value of this optimization is in not loading irrelevant blocks and not having CockroachDB iterate over key-value pairs in these irrelevant blocks.
- When Next goes past the end of a block, skip blocks that don’t overlap with the TBI interval.

**Comment by jbowens** [comment]
> Another use case that has come up is to find expired data when enforcing row-level TTLs, where (Tstart, Tend] will be in the distant past (imagine expiring 1 year old data, and doing this expiry every day, so we would be effectively looking for (now-366d, now-365d]. Since TBI filters at the granularity of sstables, and L6 contains 90% of the data, the expectation is that the current TBI will be ineffective for this use case (the finer-grained approach described here would likely be effective)

Once we have knowledge of MVCC in Pebble, we could explore compaction-time discarding of keys expired due to row-level TTLs instead. Like [you mentioned](https://github.com/cockroachdb/cockroach/issues/57260#issuecomment-739955804) it's a simpler case than general compaction-time GC. Maybe it would be a good stepping stone? MVCCStats are still an obstacle.

**Comment by sumeerbhola** [comment]
> Once we have knowledge of MVCC in Pebble, we could explore compaction-time discarding of keys expired due to row-level TTLs instead. Like you mentioned it's a simpler case than general compaction-time GC. Maybe it would be a good stepping stone? MVCCStats are still an obstacle.

Yes, that is possible, though (a) compaction-time TTL enforcement may miss things that we would need to delete by doing an efficient scan, so an efficient scan would still be useful, (b) there is (possibly) a requirement to have TTLs play nicely with CDC (normal GC is deleting obsoleted versions, while TTLs can delete the latest version), which is hard to satisfy with compaction-time TTL (there is a lot of discussion in the margins of https://docs.google.com/document/d/1aq4jft6LVZ4yb2M4v0mGHbIvTu6Xm3_bbAISFof_HG0/edit#heading=h.zgshrzthknjr)

**Comment by sumeerbhola** [comment]
> Maintain (in a time index in the sstable) what is conceptually a set containing [min,max] timestamps for each ssblock.

I am inclined towards sticking it in the usual single-level or two-level index, since this is per data block (group of data block for the two-level index) info that we already have a place for. I don't think we really need the ability to search in the index efficiently along both the key and time dimension. We would continue to binary search in the index only using the key dimension, and then linearly step forward/backward until we find a block that overlaps in time.

**Comment by sumeerbhola** [comment]
The integration on the CockroachDB side is also done. The only remaining item is https://github.com/cockroachdb/cockroach/issues/73768

**Comment by jbowens** [comment]
Have we considered also pushing down filtering of keys within blocks? Filtering by timestamp at the level of a block or sstable iterator seems like it would be a win, avoiding unnecessary full-key comparisons during merging.

**Comment by sumeerbhola** [comment]
> Have we considered also pushing down filtering of keys within blocks? Filtering by timestamp at the level of a block or sstable iterator seems like it would be a win, avoiding unnecessary full-key comparisons during merging.

I didn't consider it, but it seems worth experimenting with and benchmarking. If most data is in L6, which is where block property filters are more likely to have false positives, that creates a likelihood that the mergingIter heap will have fewer than O(number of levels) elements, and that the L6 level iterator will stay at the top of the heap, which would mean each step is 1 key comparison. That can be significant enough, if keys are long.
Do you want to create an issue?
